### ISSUE TITLE
Make tokenize recognize Other_ID_Start and Other_ID_Continue chars

### ISSUE THREAD

**Issue by 3620556a-52eb-410d-958f-a012cdc382e8** [python/cpython#68382]
BPO | [24194](https://bugs.python.org/issue24194)
--- | :---
Nosy | @terryjreedy, @meadori, @iritkatriel
Dependencies | <li>bpo-12731: python lib re uses obsolete sense of \w in full violation of UTS#18 RL1.2a</li><li>bpo-12734: Request for property support in Python re lib</li>
Files | <li>[issue24194-v0.patch](https://bugs.python.org/file42808/issue24194-v0.patch "Uploaded as text/plain at 2016-05-11.01:45:32 by @meadori")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/terryjreedy'
closed_at = None
created_at = <Date 2015-05-14.13:00:27.728>
labels = ['type-bug', 'library', '3.9', '3.10', '3.11']
title = 'Make tokenize recognize Other_ID_Start and Other_ID_Continue chars'
updated_at = <Date 2022-01-17.11:25:14.419>
user = 'https://bugs.python.org/JoshuaLandau'
```

bugs.python.org fields:
```python
activity = <Date 2022-01-17.11:25:14.419>
actor = 'vstinner'
assignee = 'terry.reedy'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2015-05-14.13:00:27.728>
creator = 'Joshua.Landau'
dependencies = ['12731', '12734']
files = ['42808']
hgrepos = []
issue_num = 24194
keywords = ['patch']
message_count = 5.0
messages = ['243188', '265286', '313851', '410718', '410731']
nosy_count = 4.0
nosy_names = ['terry.reedy', 'meador.inge', 'Joshua.Landau', 'iritkatriel']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue24194'
versions = ['Python 3.9', 'Python 3.10', 'Python 3.11']
```

</p></details>

**Comment by alexmojaki** [comment]
I recently ran into this while working with `asttokens`, hence the PR above. IPython's tracebacks rely on `asttokens`, and entering this code:

```python
In [1]: ℘℘ = 1
   ...: print(℘℘ .℘℘ )
```
crashes with a huge traceback which ends with:

```
Expected token NAME:None, got ERRORTOKEN:'℘' on line 2 col 11
```

Here's the workaround in that PR:

```python
  def patched_generate_tokens(original_tokens):
    # type: (Iterator[tokenize.TokenInfo]) -> Iterator[tokenize.TokenInfo]
    """
    Fixes tokens yielded by `tokenize.generate_tokens` to handle more non-ASCII characters in identifiers.
    Workaround for https://github.com/python/cpython/issues/68382.
    Should only be used when tokenizing a string that is known to be valid syntax,
    because it assumes that error tokens are not actually errors.
    Combines groups of consecutive NAME and/or ERRORTOKEN tokens into a single NAME token.
    """
    for key, group_iter in itertools.groupby(
        original_tokens,
        lambda t: tokenize.NAME if t.type == tokenize.ERRORTOKEN else t.type,
    ):
      group = list(group_iter)  # type: List[tokenize.TokenInfo]
      if key == tokenize.NAME and len(group) > 1 and any(tok.type == tokenize.ERRORTOKEN for tok in group):
        line = group[0].line  # type: str
        assert {tok.line for tok in group} == {line}
        yield tokenize.TokenInfo(
          type=tokenize.NAME,
          string="".join(t.string for t in group),
          start=group[0].start,
          end=group[-1].end,
          line=line,
        )
      else:
        for tok in group:
          yield tok
```

This seems to work but it's obviously a shaky hack. A proper fix would be great.

**Comment by lysnikolaou** [comment]
This has been fixed in 3.12.
